x-pack/plugins/maps/public/connected_components/mb_map/tooltip_control/tooltip_control.tsx

```diff
@@ -73,19 +75,55 @@ export class TooltipControl extends Component<Props, {}> {
     this.props.mbMap.on('mousemove', this._updateHoverTooltipState);
     this.props.mbMap.on('click', this._lockTooltip);
     this.props.mbMap.on('remove', this._setIsMapRemoved);
+    this.props.mbMap.on('sourcedata', this._onSourceData);
   }
 
   componentWillUnmount() {
     this.props.mbMap.off('mouseout', this._onMouseout);
     this.props.mbMap.off('mousemove', this._updateHoverTooltipState);
     this.props.mbMap.off('click', this._lockTooltip);
     this.props.mbMap.off('remove', this._setIsMapRemoved);
+    this.props.mbMap.off('sourcedata', this._onSourceData);
   }
 
   _setIsMapRemoved = () => {
     this._isMapRemoved = true;
   };
 
+  _onSourceData = (e: MapSourceDataEvent) => {
+    if (
+      e.sourceId &&
+      e.sourceId !== SPATIAL_FILTERS_LAYER_ID &&
+      e.dataType === 'source' &&
+      (e.source.type === 'vector' || e.source.type === 'geojson')
+    ) {
+      // map features changed, update tooltip state to match new map state.
+      this._updateOpentooltips();
+    }
+  };
+
+  _updateOpentooltips = _.debounce(() => {
+    if (this.props.openTooltips.length === 0) {
+      return;
+    }
+
+    const openTooltips = this.props.openTooltips
+      .map((tooltipState) => {
+        const mbFeatures = this._getMbFeaturesUnderPointer(
+          this.props.mbMap.project(tooltipState.location)
+        );
+        return {
+          ...tooltipState,
+          features: this._getTooltipFeatures(mbFeatures, tooltipState.isLocked, tooltipState.id),
+        };
+      })
+      .filter((tooltipState) => {
+        return tooltipState.features.length > 0;
+      });
+
+    this.props.updateOpenTooltips(openTooltips);
+  }, 300);
```

Comment:
Nit: If it's not too difficult, it would be nice to update (remove) the tooltips immediately if the layer is hidden. But not a big deal.

Reply:
I left debounce because there are so many sourcedata events. I could shorten the debounce. What do you think?

Reply:
> I left debounce because there are so many sourcedata events. I could shorten the debounce. What do you think?

Nah, I wouldn't worry about it. It's fine as is.